File: android/src/main/java/expo/modules/oseelabs/media/metadata/ExpoMediaMetadataModule.kt
package expo.modules.oseelabs.media.metadata

import android.media.MediaMetadataRetriever
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise

class ExpoMediaMetadataModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("MediaMetadata")

        Function("getMetadata") { path: String, promise: Promise ->
            val retriever = MediaMetadataRetriever()
            try {
                retriever.setDataSource(path)
                val metadata = retriever.extractMetadata(fields = MetadataField.entries)
                promise.resolve(metadata)
            } catch (e: Exception) {
                promise.reject("Error", "Could not retrieve title", e)
            } finally {
                retriever.release()
            }
        }
    }
}

File: src/ExpoMediaMetadataModule.ts
import { NativeModule, requireNativeModule } from 'expo';

declare class MediaMetadataModule extends NativeModule {
    getMetadata(path: string): Promise<Map<string, string|null>>
}

const MediaMetadata = requireNativeModule<MediaMetadataModule>('MediaMetadata');

export default class ExpoMediaMetadata {
    static async getMetadata(path: string): Promise<Map<string, string|null> | unknown> {
        try {
            const metadata = await MediaMetadata.getMetadata(path);
            return metadata;
        } catch (error: unknown) {
            console.error("Error retrieving metadata", error);
            return error;
        }
    }
}